Sentiment analysis

sophieball opened the pull request:
- Added sentiment analysis
- Added hypothesis testing for model performances
- Fixed the problem that pronoun counts were off

sophieball commented:
Build failed when collecting some python libs....

@CaptainEmerson Can you try running the current code? I output a csv for features (excluding text) in `features.csv` and print out some comment ids in the `.log`, eg
```
INFO:root:number of words above 95 quantile:
INFO:root:['302259954', '134425013', '721961187', '510276577', '509329861', '258195728', '547621451', '364692786', '751421866', '725754561', '552072224', '719750493', '434693240', '427300313', '735625724', '738128809', '478372678', '665223127', '511749880', '645802909']
INFO:root:long comments but no pronouns:
INFO:root:['317782036', '167765495', '330196930', '434348569', '427875807', '735697014', '626281783', '415712279', '705111128', '336797486', '283710495', '400543083', '661853216', '122701873', '365552246', '680245356', '478408247', '457206578', '336804781', '621880995']
```

CaptainEmerson commented:
@sophieball, main/train_classifier_g builds ok. What target did you want me to build and run?

sophieball commented:
> @sophieball, main/train_classifier_g builds ok. What target did you want me to build and run?

Please run `train_classifier_g` and there should be these files:

`features.csv`
`train_classifier_[time].log`, which contains some _ids of `number of words above 95 quantile:` and `long comments but no pronouns:`
`data_stats.csv`, `neg_label_data_stats.csv`, and `pos_label_data_stats.csv` contain descriptions of features. 

The strange thing we wanted to investigate is the number of pronouns (e.g., `1st_person` and `2nd_person`) is very low (in `data_stats.csv`). 1st person: max 3, avg 0.17, 2nd person: max 1, avg 0.06. 

For reference, in OSS reported pushback comments, 1st person pronoun count max is 21, and avg 0.56; 2nd person max 19 and avg 0.39.

sophieball commented:
I added sentiment analysis as a feature

CaptainEmerson commented:
Probably related to sentiment analysis:

```
Traceback (most recent call last):
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 139, in <module>
    trained_model = train_model(training, model_name="rf", what_data=what_data)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 39, in train_model
    s.set_train_set(training_data)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/suite.py", line 277, in set_train_set
    self.all_train_data = create_features.create_features(
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/create_features.py", line 166, in create_features
    comments_df = convo_politeness.get_politeness_score(
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/convo_politeness.py", line 215, in get_politeness_score
    scores = polite_score(corpus)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/convo_politeness.py", line 108, in polite_score
    corpus.get_utterance_ids()[0]).meta["politeness_strategies"])
IndexError: list index out of range

```

sophieball commented:
> Probably related to sentiment analysis:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 139, in <module>
>     trained_model = train_model(training, model_name="rf", what_data=what_data)
>   File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 39, in train_model
>     s.set_train_set(training_data)
>   File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/suite.py", line 277, in set_train_set
>     self.all_train_data = create_features.create_features(
>   File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/create_features.py", line 166, in create_features
>     comments_df = convo_politeness.get_politeness_score(
>   File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/convo_politeness.py", line 215, in get_politeness_score
>     scores = polite_score(corpus)
>   File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/convo_politeness.py", line 108, in polite_score
>     corpus.get_utterance_ids()[0]).meta["politeness_strategies"])
> IndexError: list index out of range
> ```

I'm not sure what the problem was because this code was before sentiment analysis. 

I think I fixed the mistakes in pronoun counts - some comments may be too short to have any politeness feature, therefore, the number of rows in the politeness results may not match that of the original data. When I concatenated them, there was a mismatch. 

I also added a t-test to compare different collections of features.

CaptainEmerson commented:
I ran from master and got the same error. Maybe something changed in the dependencies?

sophieball commented:
It could be that there's no conversation in your corpus `corpus.get_utterance_ids()[0]).meta["politeness_strategies"])` Did you change the data? Or maybe it's due to some code changes related to data? I added a `logging.info` check to see how many utterances are there.

CaptainEmerson commented:
I think that the data is the same. The log does say "INFO:root:num utterances: 0"

sophieball commented:
@CaptainEmerson I added several `logging.info` to see where did we lose all the utterances

CaptainEmerson commented:
```
Traceback (most recent call last):
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 16, in <module>
    from src import suite
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/suite.py", line 21, in <module>
    from src import create_features
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/create_features.py", line 160
    comments_df["text"] = comments_df["text"].replace(np.nan, "-")
    ^
SyntaxError: invalid syntax

```

CaptainEmerson commented:
I still get the error. Here's the log output:
INFO:root:Begin creating features, rows: 0
INFO:root:Replaced NaN, rows: 0
INFO:root:Removed empty comments, rows: 0
INFO:root:Beging calculating politeness, rows: 0
INFO:root:Num speakers: 0
INFO:root:Done preparing corpus, rows: 0
INFO:root:num utterances: 0

sophieball commented:
> I still get the error. Here's the log output:
> INFO:root:Begin creating features, rows: 0
> INFO:root:Replaced NaN, rows: 0
> INFO:root:Removed empty comments, rows: 0
> INFO:root:Beging calculating politeness, rows: 0
> INFO:root:Num speakers: 0
> INFO:root:Done preparing corpus, rows: 0
> INFO:root:num utterances: 0

Hmmmmm... I added more checks.. Maybe I messed things up when renaming columns?

CaptainEmerson commented:
New Error:
```
Traceback (most recent call last):
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 135, in <module>
    [training, unlabeled] = receive_data.receive_data(what_data)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/receive_data.py", line 26, in receive_data
    training_data["_id"] = training_data["_id"].astype(str) + "_____" + \
NameError: name 'training_data' is not defined
```

Log:
```
INFO:root:File name: Google
INFO:root:Training the model and predicting labels.
INFO:root:Reading data.
```

sophieball commented:
ooopss.. fixed before I pushed but forgot to added the fixed version :p

CaptainEmerson commented:
Back to the usual out of range error. Logs:
```
INFO:root:File name: Google
INFO:root:Training the model and predicting labels.
INFO:root:Reading data.
INFO:root:changed column names, rows: 146932
INFO:root:start training, rows: 0
INFO:root:done setup, rows: 0
INFO:root:Loading data.
INFO:root:Begin creating features, rows: 0
INFO:root:Replaced NaN, rows: 0
INFO:root:Removed empty comments, rows: 0
INFO:root:Beging calculating politeness, rows: 0
INFO:root:Num speakers: 0
INFO:root:Done preparing corpus, rows: 0
INFO:root:num utterances: 0
```

sophieball commented:
This probably solves the problem?

CaptainEmerson commented:
Yes! Progress. But new error:

```
Training classifier model..
Traceback (most recent call last):
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 136, in <module>
    trained_model = train_model(training, model_name="rf", what_data=what_data)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 40, in train_model
    s.set_train_set(training_data)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/suite.py", line 279, in set_train_set
    self.all_train_data = create_features.create_features(
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/create_features.py", line 176, in create_features
    SentiCR_model = SentiCR.SentiCR(algo="GBT",training_data= oracle_data)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/SentiCR/SentiCR.py", line 196, in __init__
    self.model = self.create_model_from_training_data()
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/SentiCR/SentiCR.py", line 228, in create_model_from_training_data
    training_comments = self.training_data["text"].map(preprocess_text).tolist()
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/deps/pypi__pandas/pandas/core/series.py", line 4160, in map
    new_values = super()._map_values(arg, na_action=na_action)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/deps/pypi__pandas/pandas/core/base.py", line 870, in _map_values
    new_values = map_f(values, mapper)
  File "pandas/_libs/lib.pyx", line 2859, in pandas._libs.lib.map_infer
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/SentiCR/SentiCR.py", line 177, in preprocess_text
    comments = handle_negation(comments)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/SentiCR/SentiCR.py", line 145, in handle_negation
    part_of_speech = nltk.tag.pos_tag(allwords,tagset='universal')
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/deps/pypi__nltk/nltk/tag/__init__.py", line 164, in pos_tag
    tagger = _get_tagger(lang)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/deps/pypi__nltk/nltk/tag/__init__.py", line 106, in _get_tagger
    tagger = PerceptronTagger()
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/deps/pypi__nltk/nltk/tag/perceptron.py", line 168, in __init__
    find("taggers/averaged_perceptron_tagger/" + PICKLE)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/deps/pypi__nltk/nltk/data.py", line 583, in find
    raise LookupError(resource_not_found)
LookupError: 
**********************************************************************
  Resource averaged_perceptron_tagger not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('averaged_perceptron_tagger')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load taggers/averaged_perceptron_tagger/averaged_perceptron_tagger.pickle

  Searched in:
    - '/usr/local/google/home/emersonm/nltk_data'
    - '/usr/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************

```

sophieball commented:
I added it to `src/download_data.py`

CaptainEmerson commented on main/feed_data.R:
```diff
@@ -7,10 +7,14 @@ library(readr)
+#df <- read.csv("src/data/pushback_survey_prs_sampled.csv")
+df <- read.csv("src/data/pushback_reported_prs_w_neg.csv")
 #df <- read.csv("src/data/issues_prs.csv")
+#df <- read.csv("src/data/random_sample_10000_prs_body_comments.csv")
```
Are the changes in this file part of what you want committed?

CaptainEmerson commented on main/feed_data.R:
```diff
@@ -23,5 +27,6 @@ system2("main/train_classifier_g",
 #system2("main/politeness_logi",
 #system2("src/convo_word_freq_diff",
         args = "prs",
+        #args = "issues",
```
ditto

CaptainEmerson reviewed (commented):
This code works for me. I've added a couple of preliminary comments here. Are there any final changes you want to make before I review this PR?

sophieball commented:
Lemme tidy up the code and make it a PR

CaptainEmerson commented on src/SentiCR/SentiCR.py:
```diff
@@ -0,0 +1,367 @@
+from __future__ import print_function
```
This directory appears to just be a fork of https://github.com/senticr/SentiCR/tree/master/SentiCR. I'd recommend:
1) Adding the LICENSE file to this directory
2) Add a README to this directory, noting that it's a fork. 
3) (Optional) If you've made modifications to this directory, the most engineering-y way to do this would be to commit the file as-is, then make the changes in a follow up PR.

sophieball replied:
oh.... I'll do 1 and 2 then. I made several changes needed for bazel

CaptainEmerson replied:
I like the bazel change. Were you going to add those files in this PR?

sophieball replied:
these files are already after bazel modification... I didn't push the original ones

CaptainEmerson commented on src/receive_data.py:
```diff
@@ -23,10 +23,13 @@ def receive_data(what_data):
   unlabeled = data.loc[data["training"] == 0].drop(["label"], axis = 1)
-  return [training, unlabeled]
+  #return [training, unlabeled]
+  return [data, data]
```
What's going on here? Why make training and unlabeled data if you're not going to use it?

sophieball replied:
This is when Naveen would run the trained classifier on some unlabeled data, but we only pass in data we have labels for.. I can make it `return [training, _]`

sophieball commented on src/suite.py:
```diff
@@ -58,6 +64,7 @@
 FLIP = 1
 DONT_FLIP = 0
 
+NUM_TRIALS = 2
```
@CaptainEmerson Can you run the `train_classifier_g` after you change this line to `NUM_TRIALS = 10`?

sophieball replied:
@CaptainEmerson  Also we need results from `convo_word_freq_diff`, but I think we've run it it and the code hasn't changed.. do you still have the result saved somewhere? Or we can just run it again - with the data cleaning you did.

CaptainEmerson replied:
I'm running it now, but last time I tried I got:

```
Traceback (most recent call last):
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 136, in <module>
    trained_model = train_model(training, model_name="rf", what_data=what_data)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/main/train_classifier_g.py", line 77, in train_model
    current_scores = s.self_issue_classification_all(model_name, fid)
  File "/usr/local/google/home/emersonm/toxicity-detector/bazel-bin/main/train_classifier_g.runfiles/__main__/src/suite.py", line 567, in self_issue_classification_all
    N_TRIALS,
NameError: name 'N_TRIALS' is not defined

```

CaptainEmerson replied:
Please change N_TRIALS to NUM_TRIALS

CaptainEmerson replied:
> @CaptainEmerson Also we need results from convo_word_freq_diff, but I think we've run it it and the code hasn't changed.. do you still have the result saved somewhere? Or we can just run it again - with the data cleaning you did.

What result files do you need? I'm also happy to run it again, but I don't see that I did any data cleaning.

sophieball replied:
I need `train_classifier_16xxxxx.log` and `fighting_words_sorted.csv`. I remember you removed some automatically  generated phrases for the fighting words task, before we showed the results to Carolyn

sophieball replied:
Another dataset is linguistic vs. non-linguistic pushback

sophieball replied:
I need the same files

sophieball commented on src/suite.py:
```diff
@@ -615,6 +615,7 @@ def self_issue_classification_all(self, model_name, fid):
             predicted_threads["prediction"].map(lambda x: int(x>0))
     predicted_thread_label = predicted_threads["thread_prediction"]
 
+    logging.info("Using after SE adjustment prediction to calculate thread accuarcy")
```
fixed

sophieball replied:
fixed

sophieball commented:
@CaptainEmerson  Can you run the new code again? The only change is that I'm printing out the scores that I used for the t-tests. I think the python `t-test` I was using is not the Student t-test we normally use, which is the `t.test` on R. I used R to run `t.test` with these scores and I think the results make more sense: on **all pushback datasets and toxic issues,** **logs-based** has the highest performance; **on toxic PRs, text-based is the best** and the combined is **not significantly** better than the text-based only.

CaptainEmerson commented:
I added two log files to the shared directory.

sophieball commented:
@CaptainEmerson Can you run the new code? Just wanted to see if 20 trials would change the t-tests

CaptainEmerson commented:
Done.
